src/dotty/tools/dotc/transform/InterceptedMethods.scala

```diff
@@ -96,14 +96,31 @@ class InterceptedMethods extends MiniPhaseTransform {
         s"that means the intercepted methods set doesn't match the code")
       tree
     }
-    lazy val Select(qual, _) = tree.fun
+    lazy val qual = tree.fun match {
+      case Select(qual, _) => qual
+      case ident @ Ident(_) =>
+        ident.tpe match {
+          case TermRef(prefix: TermRef, name) =>
+            tpd.ref(prefix)
+          case TermRef(prefix: ThisType, name) =>
+            tpd.This(prefix.cls)
+        }
+
+    }
     val Any_## = this.Any_##
     val Any_!= = defn.Any_!=
     val rewrite: Tree = tree.fun.symbol match {
       case Any_## =>
           poundPoundValue(qual)
       case Any_!= =>
-          qual.select(defn.Any_==).appliedToArgs(tree.args).select(defn.Boolean_!)
+        assert(tree.args.size == 1)
+        val lhs = qual.tpe
+        val rhs = tree.args.head.tpe
+        if (!(lhs <:< rhs || rhs <:< lhs) && rhs != defn.NullType && lhs != defn.NullType) {
```

Comment:
Here if I don't add `rhs != defn.NullType && lhs != defn.NullType` I start getting failures when comparing `java.lang.Object` some other types including types in `scala.collection`. 
@DarkDimius I wonder if this is a bug or am I testing it the wrong way?